crates/fuel-gas-price-algorithm/src/utils.rs

```diff
@@ -1,20 +1,746 @@
+    // x^y = e^(y*ln(x))
+    // we gain around 90%
+    // if we have blocks / percentage greater than m x n, we have to compute it
+    let multiple =
+        if blocks > BLOCK_COUNT_M as u32 || percentage > PERCENTAGE_COUNT_N as u64 {
+            f64::exp(blocks as f64 * (1.0 + percentage as f64 / 100.0).ln())
+        } else {
+            PRECOMPUTED_EXP[blocks as usize][percentage as usize]
+        };
+
     let mut approx = new_exec_price as f64 * multiple;
-    // account for rounding errors and take a slightly higher value
+
+    // Account for rounding errors and take a slightly higher value
+    // Around the `ROUNDING_ERROR_CUTOFF` the rounding errors will cause the estimate to be too low.
+    // We increase by `ROUNDING_ERROR_COMPENSATION` to account for this.
+    // This is an unlikely situation in practice, but we want to guarantee that the actual
+    // gas price is always equal or less than the estimate given here
     const ROUNDING_ERROR_CUTOFF: f64 = 16948547188989277.0;
-    if approx > ROUNDING_ERROR_CUTOFF {
-        const ROUNDING_ERROR_COMPENSATION: f64 = 2000.0;
-        approx += ROUNDING_ERROR_COMPENSATION;
-    }
+    const ROUNDING_ERROR_COMPENSATION: f64 = 2000.0;
+
+    // prevent branching in compiled code
+    approx +=
+        ROUNDING_ERROR_COMPENSATION * ((approx > ROUNDING_ERROR_CUTOFF) as u8 as f64);
+
     // `f64` over `u64::MAX` are cast to `u64::MAX`
     approx.ceil() as u64
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use proptest::prelude::*;
+
+    // keep the previous implementation for comparison
+    #[allow(clippy::cast_possible_truncation)]
+    fn cumulative_percentage_change_plain(
+        start_gas_price: u64,
+        best_height: u32,
+        percentage: u64,
+        target_height: u32,
+    ) -> u64 {
+        let blocks = target_height.saturating_sub(best_height) as f64;
+        let percentage_as_decimal = percentage as f64 / 100.0;
+        let multiple = (1.0f64 + percentage_as_decimal).powf(blocks);
+        let mut approx = start_gas_price as f64 * multiple;
+        // Account for rounding errors and take a slightly higher value
+        // Around the `ROUNDING_ERROR_CUTOFF` the rounding errors will cause the estimate to be too low.
+        // We increase by `ROUNDING_ERROR_COMPENSATION` to account for this.
+        // This is an unlikely situation in practice, but we want to guarantee that the actual
+        // gas price is always equal or less than the estimate given here
+        const ROUNDING_ERROR_CUTOFF: f64 = 16948547188989277.0;
+        if approx > ROUNDING_ERROR_CUTOFF {
+            const ROUNDING_ERROR_COMPENSATION: f64 = 2000.0;
+            approx += ROUNDING_ERROR_COMPENSATION;
+        }
+        // `f64` over `u64::MAX` are cast to `u64::MAX`
+        approx.ceil() as u64
+    }
+
+    proptest! {
+        #[test]
+        fn plain_and_optimized_produce_same_result(
```

Comment:
I don't think we need to maintain a test for comparing to old code. There are hundreds of possible implementations :).

Reply:
yeah, but the proptest exists to easily show what's going on ~ and to make sure the implementations match